### ISSUE TITLE
[DDP][Gradient Accumulation] Speed can be accelerated.

### ISSUE THREAD

**Issue by MagicFrogSJTU** [ultralytics/yolov5#790]
## 🚀 Feature
The training speed can be accelerated when combining DDP and gradient accumulation.
When applying gradient accumulation, the optimizer.step() is called every K steps intead of every step. And as we know every training step (with loss.backward()) DDP would activate an all_reduce. Therefore, there are K-1 steps of all_reduce wasted! Considering the time cost of all_reduce especially for multi-node training, the training speed can be accelerated a lot if we supply an option to call DDP all_reduce every K step.

## Pitch
The problem can be easily solved in 2 lines.
```python
model = DDP(model)

optimizer.zero_grad()
for i, (data, label) in enumerate(dataloader):
    # New Line 1
    my_context = model.no_sync if local_rank != -1 and i % K != 0 else nullcontext
    # New Line 2
    with my_context():
        prediction = model(data)
        loss_fn(prediction, label).backward()
    if i % K == 0:
        optimizer.step()
        optimizer.zero_grad()
```

## Additional context

The only question is whether we need to address this specail case where the user has limited GPUs but want to try experiments of big big batch size.

**Comment by glenn-jocher** [comment]
@MagicFrogSJTU wow, this is a really simple but effective change! Can you submit a PR and I'll test the branch?

EDIT: will try to compare to https://github.com/ultralytics/yolov5/pull/500#issuecomment-670194132

**Comment by glenn-jocher** [comment]
@MagicFrogSJTU based on https://stackoverflow.com/questions/3024925/create-a-with-block-on-several-context-managers I wrote this up, does this look good?

```python
            # Autocast
            reduce = model.no_sync if rank != -1 and ni % accumulate != 0 else contextlib.nullcontext()  # NEW
            with amp.autocast(enabled=cuda), reduce:  # NEW
                # Forward
                pred = model(imgs)

                # Loss
                loss, loss_items = compute_loss(pred, targets.to(device), model)  # scaled by batch_size
                if rank != -1:
                    loss *= opt.world_size  # gradient averaged between devices in DDP mode
```
